File: src/BaekJoon/java/baekjoon_1012.java
package BaekJoon.java;
// https://www.acmicpc.net/problem/1012

import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class baekjoon_1012 {
	public static class coordinate {
		int x;
		int y;
		coordinate(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}

	public static boolean checked(int[][] field, int x, int y) {
		if(x > 0 && y > 0 && x < (field.length-1) && y < (field[0].length-1))
			return true;

		return false;
	}

	public static void bfs(int[][] field, boolean[][] visited, int x, int y) {
		Queue<coordinate> q = new LinkedList<coordinate>();
		int[] dx = {-1, 0, 1, 0};
		int[] dy = {0, 1, 0, -1};
		coordinate c;

		q.add(new coordinate(x, y));

		while(!q.isEmpty()) {
			c = q.poll();

			for(int i = 0; i < dx.length; i++) {
				int currX = c.x + dx[i];
				int currY = c.y + dy[i];
				if(checked(field, currX, currY)) {
					if(field[currX][currY] == 1 && !visited[currX][currY]) {
						visited[currX][currY] = true;
						q.add(new coordinate(currX, currY));
					}
				}
			}
		} // end of while
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int test_case = sc.nextInt();
		int[][] field = null;
		boolean[][] visited = null;
		int count = 0;

		for(int i = 0; i < test_case; i++) {
			int m = sc.nextInt(); // 가로
			int n = sc.nextInt(); //; 세로
			field = new int[m+2][n+2];
			visited = new boolean[m+2][n+2];
			int k = sc.nextInt(); // 배추 개수
			count = 0;
			for(int j = 0; j < k; j++) {
				int x = sc.nextInt();
				int y = sc.nextInt();
				field[x+1][y+1] = 1;
			}

			for(int j = 1; j < field.length-1; j++) {
				for(int l = 1; l < field[i].length-1; l++) {
					if(field[j][l] == 1 && !visited[j][l]) {
						bfs(field, visited, j, l);
						count++;
					}
				}
			}
			System.out.println(count);
		}

		sc.close();
	}
}

File: src/BaekJoon/Rust/baekjoon_1012.rs
use std::collections::VecDeque;
use std::io::stdin;

struct Coordinate {
    x: i32,
    y: i32,
}

impl Coordinate {
    fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }
    
}

fn checked(field: &Vec<Vec<i32>>, x: usize, y: usize) -> bool {
    if x > 0 && y > 0 && x < field.len() - 1 && y < field[0].len() - 1 {
        true
    } else {
        false
    }
}

fn bfs(field: &Vec<Vec<i32>>, visited: &mut Vec<Vec<bool>>, x: i32, y: i32) {
    let mut queue: VecDeque<Coordinate> = VecDeque::new();
    const DX: [i32; 4] = [-1, 0, 1, 0];
    const DY: [i32; 4] = [0, 1, 0, -1];
    
    queue.push_back(Coordinate::new(x, y));
    
    while !queue.is_empty() {
        let c: Coordinate = queue.pop_front().unwrap();
        
        for index in 0..4 {
            let curr_x: usize = (c.x + DX[index]) as usize;
            let curr_y: usize = (c.y + DY[index]) as usize;
            
            if checked(field, curr_x, curr_y) && field[curr_x][curr_y] == 1 && !visited[curr_x][curr_y] {
                visited[curr_x][curr_y] = true;
                queue.push_back(Coordinate::new(curr_x as i32, curr_y as i32));
            }
        }
    }
}

pub fn baekjoon_1012() {
    let mut test_case: usize = stdin().lines().next().unwrap().unwrap().trim().parse::<usize>().unwrap();
    
    for index in 0..test_case {
        let mut line: String = String::new();
        let mut count: i32 = 0;
        
        stdin().read_line(&mut line).unwrap();
        
        let m_n_k: Vec<usize> = line.split_whitespace().map(|s| s.parse::<usize>().unwrap()).collect();
        let (m, n, k) = (m_n_k[0], m_n_k[1], m_n_k[2]);
        
        let mut field: Vec<Vec<i32>> = vec![vec![0; n + 2]; m + 2];
        let mut visited: Vec<Vec<bool>> = vec![vec![false; n + 2]; m + 2];
        
        for j in 0..k {
            let mut sub_line: String = String::new();
            stdin().read_line(&mut sub_line).unwrap();
            
            let x_y: Vec<usize> = sub_line.split_whitespace().map(|s| s.parse::<usize>().unwrap()).collect();
            let (x, y) = (x_y[0], x_y[1]);
            
            
            field[x + 1][y + 1] = 1;
        }
        
        for field_x in 1..field.len() - 1 {
            for field_y in 1..field[field_x].len() - 1 {
                if field[field_x][field_y] == 1 && !visited[field_x][field_y] {
                    bfs(&field, &mut visited, field_x as i32, field_y as i32);
                    count += 1;
                }
            }
        }
        
        println!("{}", count);
    }
}